### PR TITLE
Task retries related Web UI improvements

### PR DESCRIPTION
**Pull request by losipiuk**
## Description

Render task failures related info per stage in web UI
> Is this change a fix, improvement, new feature, refactoring, or other?

improvement

> Is this a change to the core query engine, a connector, client library, or the SPI interfaces? (be specific)

Web UI

## Documentation

(x) No documentation is needed.
( ) Sufficient documentation is included in this PR.
( ) Documentation PR is available with #prnumber.
( ) Documentation issue #issuenumber is filed, and can be handled later.

## Release notes

( ) No release notes entries required.
(x) Release notes entries required with the following suggested text:

```markdown
# Web UI
* Group information about tasks in by stage. ({issue}`12099`)
* Show aggregated statistics for failed tasks, for queries that are executed with `retry-policy` set to `TASK`. ({issue}`12099`)
```

### CODE REVIEW

**Comment by losipiuk** [comment]
image

**Comment by losipiuk** [comment]
image

**Comment by arhimondr** [comment]
With task level retries we should expect for stages to have more tasks than usual. It would be great to group task information per stage. Here is similar improvement from Presto for the reference: https://github.com/prestodb/presto/pull/13158

It would also be great to group "logical tasks" together and show number of retries and CPU time spent / wasted on a "logical task".

Additionally it would be great to show memory estimate / reservation / peak reservation for each task, largest task in a stage (on a stage level) and largest task in a query.

**Comment by losipiuk** [comment]
> With task level retries we should expect for stages to have more tasks than usual. It would be great to group task information per stage. Here is similar improvement from Presto for the reference: [prestodb/presto#13158](https://github.com/prestodb/presto/pull/13158)
> 
> It would also be great to group "logical tasks" together and show number of retries and CPU time spent / wasted on a "logical task".
> 
> Additionally it would be great to show memory estimate / reservation / peak reservation for each task, largest task in a stage (on a stage level) and largest task in a query.

Makes sense. Would you prefer to have layout with tasks grouped per stage always present. Or only when we have task level retries enabled? I think changing it to always groups tasks by stages would be less confusing to users - but tell me what you think.

**Comment by arhimondr** [comment]
>  I think changing it to always groups tasks by stages would be less confusing to users - but tell me what you think.

Yeah, I agree. I thing there's no real reason why we shouldn't always try to group them.

**Comment by losipiuk** [comment]
image

**Comment by losipiuk** [comment]
@arhimondr, @martint, @linzebing PTAL

File: core/trino-main/src/main/java/io/trino/execution/QueryInfo.java [hunk 1]
```diff
@@ -79,6 +79,7 @@
     private final boolean completeInfo;
     private final Optional<ResourceGroupId> resourceGroupId;
     private final Optional<QueryType> queryType;
+    private final String retryPolicy;
```
Comment (arhimondr) [inline core/trino-main/src/main/java/io/trino/execution/QueryInfo.java:82]: Why not the enum?

**Review by arhimondr — COMMENTED** [review]
Looks good to me so far.

Do you think it would make sense to also update the home page (with the list of queries running) to include number of failed tasks (and maybe wasted CPU time?) for each query to make it easier to identify queries that experienced a task failure?

**Review by linzebing — APPROVED** [review]

File: core/trino-main/src/main/java/io/trino/execution/QueryInfo.java (hunk 1 above)
Comment (losipiuk) [reply core/trino-main/src/main/java/io/trino/execution/QueryInfo.java:82]: IDK - I think I for some reason thought that QueryInfo is in SPI and you cannot do that. WIll change.

**Comment by losipiuk** [comment]
> Do you think it would make sense to also update the home page (with the list of queries running) to include number of failed tasks (and maybe wasted CPU time?) for each query to make it easier to identify queries that experienced a task failure?

Good idea. I would go with just failed tasks. If that metrics goes up you probably want to click the query anyway and look at other, more detailed stats (also I have no clue what icon to use for wasted CPU - the icons we have so far are not very easy to comprehend already ;) )

**Comment by losipiuk** [comment]
image

**Review by arhimondr — APPROVED** [review]

**Pull request merged**